Namespace as resource fragment results in NullPointerException

### Description
<!--
Thank you for trying out Eclipse JKube!

Please fill out the following information to help us in analyzing the issue, but feel free to skip it if you don't have the information at hand or if it does not apply. Please remove every section which does not apply to the issue.
-->

We have DefaultNamespaceEnricher which seems to be taking care of creating a namespace on the fly when provided as a resource fragment. However, when I add a simple `Namespace` resource  fragment in `src/main/jkube` like this:
```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: test-jkube
```

It results in `NullPointerException`: 
```
[INFO] Building Eclipse JKube :: Quickstarts :: Maven :: External Resources 1.0.2
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- kubernetes-maven-plugin:1.0.2:resource (default-cli) @ external-resources ---
[INFO] k8s: Running generator spring-boot
[INFO] k8s: spring-boot: Using Docker image quay.io/jkube/jkube-java-binary-s2i:0.0.8 as base / builder                                                                                               
[INFO] k8s: Using resource templates from /home/rohaan/work/repos/eclipse-jkube-sample-with-resource-fragments/src/main/jkube
[INFO] k8s: jkube-controller: Adding a default Deployment
[INFO] k8s: jkube-healthcheck-spring-boot: Adding readiness probe on port 8080, path='/health', scheme='HTTP', with initial delay 10 seconds
[INFO] k8s: jkube-healthcheck-spring-boot: Adding liveness probe on port 8080, path='/health', scheme='HTTP', with initial delay 180 seconds
[INFO] k8s: jkube-revision-history: Adding revision history limit to 2
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.901 s
[INFO] Finished at: 2020-12-03T19:29:48+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.eclipse.jkube:kubernetes-maven-plugin:1.0.2:resource (default-cli) on project external-resources: Execution default-cli of goal org.eclipse.jkube:kubernetes-maven-plugin:1.0.2:resource failed.: NullPointerException -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginExecutionException
```

### Info

* Eclipse JKube version : 1.0.2
* Maven version (`mvn -v`) :
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /opt/apache-maven-3.6.3
Java version: 1.8.0_271, vendor: Oracle Corporation, runtime: /usr/lib/java/jdk1.8.0_271/jre
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.9.10-200.fc33.x86_64", arch: "amd64", family: "unix"

```
* Kubernetes / Red Hat OpenShift setup and version : Kubernetes 1.19.4

* If it's a bug, how to reproduce : Just add namespace fragment and issue `mvn k8s:resource`

* If it's a feature request, what is your use case :

* Sample Reproducer Project : https://github.com/r0haaaan/eclipse-jkube-sample-with-resource-fragments/tree/namespace-fragment

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -35,6 +35,7 @@ Usage:
 * Fix #181: Refactor PortForwardService to use Kubernetes Client Port Forwarding instead of kubectl binary
 * Fix #535: Bump JKube base images to 0.0.9
 * Fix #509: Port of ServiceDiscoveryEnricher from FMP
+* Fix #511: Namespace as resource fragment results in NullPointerException
 
 ### 1.0.2 (2020-10-30)
 * Fix #429: Added quickstart for Micronaut framework

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/ApplyService.java
@@ -76,6 +76,7 @@ import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.stream.Collectors;
 
 import static org.eclipse.jkube.kit.common.util.KubernetesHelper.getCustomResourcesFileToNameMap;
 import static org.eclipse.jkube.kit.common.util.KubernetesHelper.getKind;
@@ -1381,7 +1382,7 @@ public class ApplyService {
         this.rollingUpgradePreserveScale = rollingUpgradePreserveScale;
     }
 
-    public void processCustomEntities(List<String> customResourceDefinitions, File resourceDir, String environment, List<String> remotes) throws Exception {
+    public void applyCustomEntities(List<String> customResourceDefinitions, File resourceDir, String environment, List<String> remotes) throws Exception {
         if(customResourceDefinitions == null)
             return;
 
@@ -1398,32 +1399,14 @@ public class ApplyService {
         }
     }
 
-
-    /**
-     * Apply Namespace or Project in Kubernetes resource list if present. We need to process
-     * Namespace before processing other resources since namespace might be required while
-     * applying other resources.
-     *
-     * @param entities Collection of Kubernetes resource entities provided
-     */
-    public void applyNamespaceOrProjectIfPresent(Collection<HasMetadata> entities) {
-        entities.stream()
-                .filter(e -> e instanceof Namespace || e instanceof Project)
-                .forEach(this::applyProjectOrNamespace);
-
-        entities.removeIf(e -> e instanceof Project || e instanceof Namespace);
-    }
-
     public void applyEntities(String fileName, Set<HasMetadata> entities, KitLogger serviceLogger,
                                  long serviceUrlWaitTimeSeconds, ResourceConfig resources, File resourceDir,
                                  String environment) throws Exception {
-        // Apply Namespace/Project first
-        applyNamespaceOrProjectIfPresent(entities);
-        // Apply all items
-        applyStandardEntities(fileName, entities);
-        logExposeServiceUrl(entities, serviceLogger, serviceUrlWaitTimeSeconds);
-        processCustomEntities(resources != null ? resources.getCustomResourceDefinitions() : null,
+
+        applyStandardEntities(fileName, getK8sListWithNamespaceFirst(entities));
+        applyCustomEntities(resources != null ? resources.getCustomResourceDefinitions() : null,
                 resourceDir, environment, resources != null ? resources.getRemotes() : null);
+        logExposeServiceUrl(entities, serviceLogger, serviceUrlWaitTimeSeconds);
     }
 
     private void logExposeServiceUrl(Set<HasMetadata> entities, KitLogger serviceLogger, long serviceUrlWaitTimeSeconds) throws InterruptedException {
@@ -1434,7 +1417,7 @@ public class ApplyService {
     }
 
 
-    private void applyStandardEntities(String fileName, Set<HasMetadata> entities) {
+    private void applyStandardEntities(String fileName, List<HasMetadata> entities) {
         for (HasMetadata entity : entities) {
             if (entity instanceof Pod) {
                 Pod pod = (Pod) entity;
@@ -1448,16 +1431,6 @@ public class ApplyService {
             } else if (entity != null) {
                 apply(entity, fileName);
             }
-        }
-    }
-
-    private void applyProjectOrNamespace(HasMetadata projectOrNs) {
-        if (projectOrNs instanceof Namespace) {
-            Namespace ns = (Namespace) projectOrNs;
-            applyNamespace(ns);
-        } else if (projectOrNs instanceof Project) {
-            Project project = (Project) projectOrNs;
-            applyProject(project);
         }
     }
 
@@ -1476,5 +1449,20 @@ public class ApplyService {
             return clusterAccess.getNamespace();
         }
         return namespaceInProvidedKubernetesList;
+    }
+
+    public static List<HasMetadata> getK8sListWithNamespaceFirst(Collection<HasMetadata> k8sList) {
+        return k8sList.stream().sorted((k1, k2) -> {
+            if (isNamespaceOrProject(k1)) {
+                return -1;
+            } else if (isNamespaceOrProject(k2)) {
+                return 1;
+            }
+            return 0;
+        }).collect(Collectors.toList());
+    }
+
+    private static boolean isNamespaceOrProject(HasMetadata h) {
+        return h instanceof Namespace || h instanceof Project;
     }
 }

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployService.java
@@ -68,6 +68,7 @@ public class KubernetesUndeployService implements UndeployService {
     final String currentNamespace = currentNamespace(entities);
     undeployResources(currentNamespace, entities);
     undeployCustomResources(currentNamespace, resourceDir, resourceConfig);
+    undeployNamespace(entities);
   }
 
   private void undeployResources(String namespace, List<HasMetadata> entities) {
@@ -78,6 +79,13 @@ public class KubernetesUndeployService implements UndeployService {
         .forEach(resourceDeleter);
   }
 
+  private void undeployNamespace(List<HasMetadata> entities) {
+    final Consumer<HasMetadata> resourceDeleter = clusterScopedResourceDeleter();
+    entities.stream()
+            .filter(e -> (e instanceof Namespace || e instanceof Project))
+            .forEach(resourceDeleter);
+  }
+
   protected Consumer<HasMetadata> resourceDeleter(String namespace) {
     return resource -> {
       logger.info("Deleting resource %s %s/%s", KubernetesHelper.getKind(resource), namespace, KubernetesHelper.getName(resource));
@@ -85,6 +93,15 @@ public class KubernetesUndeployService implements UndeployService {
           .inNamespace(namespace)
           .withPropagationPolicy(DeletionPropagation.BACKGROUND)
           .delete();
+    };
+  }
+
+  protected Consumer<HasMetadata> clusterScopedResourceDeleter() {
+    return resource -> {
+      logger.info("Deleting resource %s %s", KubernetesHelper.getKind(resource), KubernetesHelper.getName(resource));
+      jKubeServiceHub.getClient().resource(resource)
+              .withPropagationPolicy(DeletionPropagation.BACKGROUND)
+              .delete();
     };
   }
 

--- a/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployService.java
+++ b/jkube-kit/config/service/src/main/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployService.java
@@ -39,6 +39,7 @@ import io.fabric8.openshift.api.model.Project;
 import static org.eclipse.jkube.kit.common.util.KubernetesHelper.getCustomResourcesFileToNameMap;
 import static org.eclipse.jkube.kit.common.util.KubernetesHelper.loadResources;
 import static org.eclipse.jkube.kit.common.util.KubernetesHelper.unmarshalCustomResourceFile;
+import static org.eclipse.jkube.kit.config.service.ApplyService.getK8sListWithNamespaceFirst;
 import static org.eclipse.jkube.kit.config.service.kubernetes.KubernetesClientUtil.getCustomResourceDefinitionContext;
 
 public class KubernetesUndeployService implements UndeployService {
@@ -56,7 +57,7 @@ public class KubernetesUndeployService implements UndeployService {
     final List<File> manifests = Stream.of(manifestFiles)
         .filter(Objects::nonNull).filter(File::exists).filter(File::isFile)
         .collect(Collectors.toList());
-    final List<HasMetadata> entities = new ArrayList<>();
+    List<HasMetadata> entities = new ArrayList<>();
     for (File manifest : manifests) {
       entities.addAll(loadResources(manifest));
     }
@@ -64,26 +65,16 @@ public class KubernetesUndeployService implements UndeployService {
       logger.warn("No such generated manifests found for this project, ignoring.");
       return;
     }
-    Collections.reverse(entities);
-    final String currentNamespace = currentNamespace(entities);
-    undeployResources(currentNamespace, entities);
+    List<HasMetadata> undeployEntities = getK8sListWithNamespaceFirst(entities);
+    Collections.reverse(undeployEntities);
+    final String currentNamespace = currentNamespace(undeployEntities);
     undeployCustomResources(currentNamespace, resourceDir, resourceConfig);
-    undeployNamespace(entities);
+    undeployResources(currentNamespace, undeployEntities);
   }
 
   private void undeployResources(String namespace, List<HasMetadata> entities) {
     final Consumer<HasMetadata> resourceDeleter = resourceDeleter(namespace);
-    entities.stream()
-        .filter(e -> !(e instanceof Namespace))
-        .filter(e -> !(e instanceof Project))
-        .forEach(resourceDeleter);
-  }
-
-  private void undeployNamespace(List<HasMetadata> entities) {
-    final Consumer<HasMetadata> resourceDeleter = clusterScopedResourceDeleter();
-    entities.stream()
-            .filter(e -> (e instanceof Namespace || e instanceof Project))
-            .forEach(resourceDeleter);
+    entities.forEach(resourceDeleter);
   }
 
   protected Consumer<HasMetadata> resourceDeleter(String namespace) {
@@ -93,15 +84,6 @@ public class KubernetesUndeployService implements UndeployService {
           .inNamespace(namespace)
           .withPropagationPolicy(DeletionPropagation.BACKGROUND)
           .delete();
-    };
-  }
-
-  protected Consumer<HasMetadata> clusterScopedResourceDeleter() {
-    return resource -> {
-      logger.info("Deleting resource %s %s", KubernetesHelper.getKind(resource), KubernetesHelper.getName(resource));
-      jKubeServiceHub.getClient().resource(resource)
-              .withPropagationPolicy(DeletionPropagation.BACKGROUND)
-              .delete();
     };
   }
 

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/ApplyServiceTest.java
@@ -14,13 +14,13 @@
 
 package org.eclipse.jkube.kit.config.service;
 
+import io.fabric8.kubernetes.api.model.Namespace;
+import io.fabric8.kubernetes.api.model.NamespaceBuilder;
 import io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinition;
 import io.fabric8.kubernetes.api.model.apiextensions.v1beta1.CustomResourceDefinitionBuilder;
 import io.fabric8.kubernetes.api.model.ConfigMap;
 import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
 import io.fabric8.kubernetes.api.model.HasMetadata;
-import io.fabric8.kubernetes.api.model.Namespace;
-import io.fabric8.kubernetes.api.model.NamespaceBuilder;
 import io.fabric8.kubernetes.api.model.Pod;
 import io.fabric8.kubernetes.api.model.PodBuilder;
 import io.fabric8.kubernetes.api.model.ReplicationController;
@@ -29,9 +29,9 @@ import io.fabric8.kubernetes.api.model.Service;
 import io.fabric8.kubernetes.api.model.ServiceBuilder;
 import io.fabric8.kubernetes.api.model.apps.Deployment;
 import io.fabric8.kubernetes.api.model.apps.DeploymentBuilder;
+import io.fabric8.openshift.api.model.DeploymentConfigBuilder;
 import io.fabric8.openshift.api.model.Project;
 import io.fabric8.openshift.api.model.ProjectBuilder;
-import io.fabric8.openshift.api.model.ProjectListBuilder;
 import io.fabric8.openshift.api.model.Route;
 import io.fabric8.openshift.api.model.RouteBuilder;
 import io.fabric8.openshift.client.server.mock.OpenShiftServer;
@@ -55,6 +55,8 @@ import java.util.HashSet;
 import java.util.Set;
 
 import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
 
 public class ApplyServiceTest {
 
@@ -70,35 +72,6 @@ public class ApplyServiceTest {
     public void setUp() {
         applyService = new ApplyService(mockServer.getOpenshiftClient(), log);
         applyService.setNamespace("default");
-    }
-
-    @Test
-    public void testApplyNamespaceOrProjectIfPresent() {
-        // Given
-        Set<HasMetadata> entities = new HashSet<>();
-        Namespace namespace = new NamespaceBuilder().withNewMetadata().withName("ns1").endMetadata().build();
-        Project project = new ProjectBuilder().withNewMetadata().withName("p1").endMetadata().build();
-        entities.add(namespace);
-        entities.add(project);
-        WebServerEventCollector collector = new WebServerEventCollector();
-        mockServer.expect().post()
-                .withPath("/api/v1/namespaces")
-                .andReply(collector.record("new-namespace").andReturn(HTTP_CREATED, namespace))
-                .once();
-        mockServer.expect().get()
-                .withPath("/apis/project.openshift.io/v1/projects")
-                .andReply(collector.record("get-project").andReturn(HTTP_OK, new ProjectListBuilder().build()))
-                .once();
-        mockServer.expect().post()
-                .withPath("/apis/project.openshift.io/v1/projectrequests")
-                .andReply(collector.record("new-projectrequest").andReturn(HTTP_CREATED, project))
-                .once();
-
-        // When
-        applyService.applyNamespaceOrProjectIfPresent(entities);
-
-        // Then
-        collector.assertEventsRecordedInOrder( "get-project", "new-projectrequest", "new-namespace");
     }
 
     @Test
@@ -253,7 +226,7 @@ public class ApplyServiceTest {
                 .once();
 
         // When
-        applyService.processCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
+        applyService.applyCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
 
         // Then
         collector.assertEventsRecordedInOrder("get-crd-virtualservice", "get-crd-gateway", "post-cr-virtualservice", "post-cr-gateway");
@@ -294,7 +267,7 @@ public class ApplyServiceTest {
                 .once();
 
         // When
-        applyService.processCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
+        applyService.applyCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
 
         // Then
         collector.assertEventsRecordedInOrder("get-crd-virtualservice", "get-crd-gateway", "get-cr-virtualservice", "put-cr-virtualservice", "get-cr-gateway", "put-cr-gateway");
@@ -336,7 +309,7 @@ public class ApplyServiceTest {
         applyService.setRecreateMode(true);
 
         // When
-        applyService.processCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
+        applyService.applyCustomEntities(crdNames, crFragmentDir, null, Collections.emptyList());
         applyService.setRecreateMode(false);
 
         // Then
@@ -351,10 +324,46 @@ public class ApplyServiceTest {
         File crFragmentDir = new File(getClass().getResource("/gateway-cr.yml").getFile()).getParentFile();
 
         // When
-        applyService.processCustomEntities(null, crFragmentDir, null, Collections.emptyList());
+        applyService.applyCustomEntities(null, crFragmentDir, null, Collections.emptyList());
 
         // Then
         assertEquals(1, mockServer.getMockServer().getRequestCount()); // This one is just /apis check
+    }
+
+    @Test
+    public void testGetK8sListWithNamespaceFirstWithNamespace() {
+        // Given
+        List<HasMetadata> k8sList = new ArrayList<>();
+        k8sList.add(new PodBuilder().withNewMetadata().withNamespace("p1").endMetadata().build());
+        k8sList.add(new ConfigMapBuilder().withNewMetadata().withName("c1").endMetadata().build());
+        k8sList.add(new NamespaceBuilder().withNewMetadata().withName("n1").endMetadata().build());
+        k8sList.add(new DeploymentBuilder().withNewMetadata().withName("d1").endMetadata().build());
+
+        // When
+        List<HasMetadata> result = ApplyService.getK8sListWithNamespaceFirst(k8sList);
+
+        // Then
+        assertNotNull(result);
+        assertEquals(k8sList.size(), result.size());
+        assertTrue(result.get(0) instanceof Namespace);
+    }
+
+    @Test
+    public void testGetK8sListWithNamespaceFirstWithProject() {
+        // Given
+        List<HasMetadata> k8sList = new ArrayList<>();
+        k8sList.add(new PodBuilder().withNewMetadata().withNamespace("p1").endMetadata().build());
+        k8sList.add(new ConfigMapBuilder().withNewMetadata().withName("c1").endMetadata().build());
+        k8sList.add(new DeploymentConfigBuilder().withNewMetadata().withName("d1").endMetadata().build());
+        k8sList.add(new ProjectBuilder().withNewMetadata().withName("project1").endMetadata().build());
+
+        // When
+        List<HasMetadata> result = ApplyService.getK8sListWithNamespaceFirst(k8sList);
+
+        // Then
+        assertNotNull(result);
+        assertEquals(k8sList.size(), result.size());
+        assertTrue(result.get(0) instanceof Project);
     }
 
     private Route buildRoute() {

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployServiceTest.java
@@ -80,7 +80,7 @@ public class KubernetesUndeployServiceTest {
   }
 
   @Test
-  public void undeployWithManifestShouldDeleteApplicableEntities(@Mocked File file) throws Exception {
+  public void undeployWithManifestShouldDeleteAllEntities(@Mocked File file) throws Exception {
     // Given
     final Namespace namespace = new NamespaceBuilder().withNewMetadata().withName("default").endMetadata().build();
     final Pod pod = new PodBuilder().withNewMetadata().withName("MrPoddington").endMetadata().build();
@@ -113,7 +113,7 @@ public class KubernetesUndeployServiceTest {
   }
 
   @Test
-  public void undeployWithManifestAndCustomResourcesShouldDeleteApplicableEntities(
+  public void undeployWithManifestAndCustomResourcesShouldDeleteAllEntities(
       @Mocked ResourceConfig resourceConfig) throws Exception {
     // Given
     final File manifest = temporaryFolder.newFile("temp.yml");

--- a/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployServiceTest.java
+++ b/jkube-kit/config/service/src/test/java/org/eclipse/jkube/kit/config/service/kubernetes/KubernetesUndeployServiceTest.java
@@ -98,12 +98,15 @@ public class KubernetesUndeployServiceTest {
     // Then
     // @formatter:off
     new Verifications() {{
-      kubernetesHelper.getKind((HasMetadata)any); times = 2;
+      kubernetesHelper.getKind((HasMetadata)any); times = 3;
       jKubeServiceHub.getClient().resource(pod).inNamespace("default")
           .withPropagationPolicy(DeletionPropagation.BACKGROUND).delete();
       times = 1;
       jKubeServiceHub.getClient().resource(service).inNamespace("default")
           .withPropagationPolicy(DeletionPropagation.BACKGROUND).delete();
+      times = 1;
+      jKubeServiceHub.getClient().resource(namespace)
+              .withPropagationPolicy(DeletionPropagation.BACKGROUND).delete();
       times = 1;
     }};
     // @formatter:on

--- a/jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/DefaultNamespaceEnricher.java
+++ b/jkube-kit/enricher/generic/src/main/java/org/eclipse/jkube/enricher/generic/DefaultNamespaceEnricher.java
@@ -18,10 +18,12 @@ import io.fabric8.kubernetes.api.model.HasMetadata;
 import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
 import io.fabric8.kubernetes.api.model.Namespace;
 import io.fabric8.kubernetes.api.model.NamespaceBuilder;
+import io.fabric8.kubernetes.api.model.NamespaceStatus;
 import io.fabric8.kubernetes.api.model.ObjectMeta;
 import io.fabric8.kubernetes.api.model.ObjectMetaBuilder;
 import io.fabric8.openshift.api.model.Project;
 import io.fabric8.openshift.api.model.ProjectBuilder;
+import io.fabric8.openshift.api.model.ProjectStatus;
 import lombok.AllArgsConstructor;
 import lombok.Getter;
 import org.eclipse.jkube.kit.common.Configs;
@@ -134,18 +136,26 @@ public class DefaultNamespaceEnricher extends BaseEnricher {
         builder.accept(new TypedVisitor<NamespaceBuilder>() {
             @Override
             public void visit(NamespaceBuilder builder) {
-                if (builder.buildStatus().getPhase().equals("active")) {
+                // Set status as empty
+                NamespaceStatus status = builder.buildStatus();
+                if (status != null && status.getPhase().equals("active")) {
                     builder.editOrNewStatus().endStatus().build();
                 }
+                // Set metadata.namespace as null
+                builder.editOrNewMetadata().withNamespace(null).endMetadata();
             }
         });
 
         builder.accept(new TypedVisitor<ProjectBuilder>() {
             @Override
             public void visit(ProjectBuilder builder) {
-                if (builder.buildStatus().getPhase().equals("active")) {
+                // Set status as empty
+                ProjectStatus status = builder.buildStatus();
+                if (status != null && status.getPhase().equals("active")) {
                     builder.editOrNewStatus().endStatus().build();
                 }
+                // Set metadata.namespace as null
+                builder.editOrNewMetadata().withNamespace(null).endMetadata();
             }
         });
     }

--- a/jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/DefaultNamespaceEnricherTest.java
+++ b/jkube-kit/enricher/generic/src/test/java/org/eclipse/jkube/enricher/generic/DefaultNamespaceEnricherTest.java
@@ -13,11 +13,13 @@
  */
 package org.eclipse.jkube.enricher.generic;
 
-import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
 import io.fabric8.kubernetes.api.model.KubernetesListBuilder;
 import io.fabric8.kubernetes.api.model.Namespace;
 import io.fabric8.kubernetes.api.model.NamespaceBuilder;
+import io.fabric8.kubernetes.api.model.apps.Deployment;
+import io.fabric8.kubernetes.api.model.apps.DeploymentBuilder;
 import io.fabric8.openshift.api.model.Project;
+import io.fabric8.openshift.api.model.ProjectBuilder;
 import mockit.Expectations;
 import mockit.Mocked;
 import org.eclipse.jkube.kit.config.resource.PlatformMode;
@@ -30,6 +32,7 @@ import java.util.Collections;
 import java.util.Properties;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.Assert.assertNull;
 
 public class DefaultNamespaceEnricherTest {
 
@@ -108,18 +111,71 @@ public class DefaultNamespaceEnricherTest {
   }
 
   @Test
-  public void enrichWithPropertiesInKubernetesShouldAddNamespace() {
+  public void enrichWithPropertiesInKubernetesShouldAddNamespaceWithStatus() {
     // Given
     properties.put("jkube.enricher.jkube-namespace.namespace", "example");
     final KubernetesListBuilder klb = new KubernetesListBuilder();
-    klb.addToItems(new NamespaceBuilder()
-        .editOrNewMetadata().withName("name").withNamespace("to-be-overwritten").endMetadata()
-        .editOrNewStatus().withPhase("active").endStatus().build());
+    Namespace namespace = new NamespaceBuilder()
+            .editOrNewMetadata().withName("name").withNamespace("to-be-overwritten").endMetadata()
+            .editOrNewStatus().withPhase("active").endStatus().build();
+    Deployment deployment = new DeploymentBuilder().withNewMetadata().withName("d1").endMetadata().build();
+    klb.addToItems(namespace, deployment);
     // When
     new DefaultNamespaceEnricher(context).enrich(PlatformMode.kubernetes, klb);
     // Then
+    assertThat(klb.build().getItems()).hasSize(2);
+    assertThat(klb.build().getItems().get(1))
+        .hasFieldOrPropertyWithValue("metadata.namespace", "example");
+  }
+
+  @Test
+  public void enrichWithPropertiesInKubernetesShouldAddProjectWithStatus() {
+    // Given
+    final KubernetesListBuilder klb = new KubernetesListBuilder();
+    klb.addToItems(new ProjectBuilder()
+            .withNewMetadata().withName("name").endMetadata()
+            .withNewStatus().withPhase("active").endStatus().build());
+    // When
+    new DefaultNamespaceEnricher(context).enrich(PlatformMode.openshift, klb);
+    // Then
     assertThat(klb.build().getItems()).hasSize(1);
     assertThat(klb.build().getItems().iterator().next())
-        .hasFieldOrPropertyWithValue("metadata.namespace", "example");
+            .hasFieldOrPropertyWithValue("metadata.namespace", null);
+  }
+
+  @Test
+  public void enrichWithNamespaceFragmentWithNoStatus() {
+    // Given
+    final KubernetesListBuilder kubernetesListBuilder = new KubernetesListBuilder();
+    kubernetesListBuilder.addToItems(new NamespaceBuilder()
+            .withNewMetadata().withName("test-jkube").endMetadata()
+            .build());
+
+    // When
+    new DefaultNamespaceEnricher(context).enrich(PlatformMode.kubernetes, kubernetesListBuilder);
+
+    // Then
+    assertThat(kubernetesListBuilder.build().getItems()).hasSize(1);
+    assertThat(kubernetesListBuilder.build().getItems().iterator().next())
+            .hasFieldOrPropertyWithValue("metadata.name", "test-jkube");
+    assertNull(kubernetesListBuilder.build().getItems().get(0).getMetadata().getNamespace());
+  }
+
+  @Test
+  public void enrichWithOpenShiftProjectFragmentWithNoStatus() {
+    // Given
+    final KubernetesListBuilder kubernetesListBuilder = new KubernetesListBuilder();
+    kubernetesListBuilder.addToItems(new ProjectBuilder()
+            .withNewMetadata().withName("test-jkube").endMetadata()
+            .build());
+
+    // When
+    new DefaultNamespaceEnricher(context).enrich(PlatformMode.openshift, kubernetesListBuilder);
+
+    // Then
+    assertThat(kubernetesListBuilder.build().getItems()).hasSize(1);
+    assertThat(kubernetesListBuilder.build().getItems().iterator().next())
+            .hasFieldOrPropertyWithValue("metadata.name", "test-jkube");
+    assertNull(kubernetesListBuilder.build().getItems().get(0).getMetadata().getNamespace());
   }
 }

